docs/plugins/asg_rollout.md

```diff
@@ -0,0 +1,85 @@
+# Auto-Scaling Group LaunchConfig Rollout Plugin
+
+## Overview
+
+This plugin adds support for rolling out changes to Auto Scaling
+Groups that have happened after a stack update. It supports various
+pre- and post-actions on the instances (using Moonshot's native SSH
+support).
+
+## Example
+
+The ASGRollout class is intended to be used within a method in your
+project's derived class of the Moonshot::CLI class.
+
+```ruby
+#!/usr/bin/env ruby
+
+require 'moonshot'
+
+class MyService < Moonshot::CLI
+  # .. normal configuration ..
+  
+  desc :asg_rollout, 'Update instances in the Auto Scaling Group'
+  def asg_rollout
+    Moonshot::Plugins::ASGRollout.new(
+      controller: controller,
+      logical_id: 'APIAutoScalingGroup',
+      # Trigger the worker process to shut down cleanly.
+      pre_detach: -> {
+        instance.exec('systemctl stop my-worker-1')
+      }
+
+      # Ensure that the worker process has completed it's work before terminating.
+      terminate_when: -> {
+        0 == instance.exec('pgrep -f my-worker-1').exitstatus
+      }
+    ).run!
+  end
+end
+
+MyService.start
+```
+
+## Configuration
+
+The ASGRollout object accepts two required keyword arguments:
+  - **controller**: A Moonshot::Controller. If you are using
+    Moonshot::CLI the `controller` method is most often when you want
+    to use.
+  - **logical_id**: The Logical resource ID from the CloudFormation
+    stack to operate on.
+
+By default, the ASGRollout class will build a list of instances with
+non-conforming LaunchConfiguration, then one at a time perform the
+following actions:
+  1. Increase the *Max* and *Desired* sizes of the Auto Scaling Group by one.
```

Comment:
`Max` will be **implicitly** increased if `Max == Desired` and you increase `Desired`, same applies for `Min` (but the other way obviously). We must pay attention to restore these values to the original state, not just decrease them by one at the last step.

Eg.: 
- let's say you have `Desired: 1` and `Max: 1`
- you first increase the desired value, you end up with `Desired: 2` and `Max: 2`
- _then_ you increase max: `Desired: 2` and `Max: 3`
- decrease both at the end, in arbitrary order: `Desired: 1` and `Max: 2`